Only show food related questions if someone has food needs

brucebolt opened the pull request:
What
----

- Changes the order of questions, so the "do you have any special dietary requirements" and "can you carry a box inside" questions come immediately after the "do you have a way of getting essential supplies delivered at the moment" question.
- Only shows the two later food questions if the user does not have a way of getting food.
- Updates the "previous" link target for subsequent food pages based on the user's answers.
- Removes the skippable questions from the "check your answers" page when the user has not answered them.
- Ensures a user is shown the skippable questions if they change their answer to say they do have problems getting food.

Not to be merged until approved by DWP call centre, and the data processing team.

How to review
-------------

- Review code changes.
- Run app locally, or use preview, to examine behaviour of `/essential-supplies` and subsequent questions when the different responses are given.  Also review the behaviour of changing your answer to the food questions.

Links
-----

Trello card: https://trello.com/c/IWPGeYTp

brucebolt committed:
Change ordering of food parcel questions

Move "do you have any dietary requirements" and "can you carry a box
inside" questions to immediately follow the "do you have a way of
getting essential supplies" question.

brucebolt committed:
Do not ask later food questions if able to get food

If a user says they are able to get essential supplies delivered, then
we don't need to ask whether they have dietary requirements or are able
to carry a box inside.

Therefore we should skip these two questions if the user can access
essential supplies.

brucebolt committed:
Allow users to reanswer food questions

There are two scenarios:

- User has intially said "yes" to being able to get essential supplies,
  so they won't have seen the dietary requirements or carry supplies
  question.  If they change their answer to "no", we need to show them
  the two later food questions in all cases.
- User has initially said "no" to being able to get essential supplies,
  so they have already seen the later two food questions.  We should
  therefore not show the later two questions if they edit their answer
  to being able to get essential supplies.

brucebolt committed:
Do not show skippable food questions if not answered

We should not be showing questions on the check your answers page where
users have not answered the question.

leenagupte commented on app/controllers/coronavirus_form/carry_supplies_controller.rb:
```diff
@@ -18,9 +18,12 @@ def submit
         format.html { render controller_path, status: :unprocessable_entity }
       end
+    elsif session[:check_answers_seen]
+      session[:essential_supplies] = @form_responses[:essential_supplies]
```
Should this be setting the value for `carry_supplies`?

leenagupte commented on app/controllers/coronavirus_form/basic_care_needs_controller.rb:
```diff
@@ -27,6 +27,6 @@ def submit
 
   def previous_path
-    carry_supplies_path
+    @form_responses[:essential_supplies] == I18n.t("coronavirus_form.questions.essential_supplies.options.option_no.label") ? carry_supplies_path : essential_supplies_path
```
This line is really long. Perhaps it could be split up? 
e.g.

```
return carry_supplies_path if @form_responses[:essential_supplies] == I18n.t("coronavirus_form.questions.essential_supplies.options.option_no.label")

essential_supplies_path
```

leenagupte commented on app/controllers/coronavirus_form/carry_supplies_controller.rb:
```diff
@@ -19,7 +19,7 @@ def submit
       end
     elsif session[:check_answers_seen]
-      session[:essential_supplies] = @form_responses[:essential_supplies]
+      session[:carry_supplies] = @form_responses[:carry_supplies]
```
Ah! You spotted it. This should probably be fixed up to the previous commit.

brucebolt replied:
Looks like I had fixed this in the wrong commit 🙊

leenagupte commented on spec/controllers/coronavirus_form/essential_supplies_controller_spec.rb:
```diff
@@ -35,11 +35,22 @@
+    it "clears session values for later food questions for a yes response" do
+      post :submit, params: { essential_supplies: I18n.t("coronavirus_form.questions.essential_supplies.options.option_yes.label") }
+      expect(session[:dietary_requirements]).to be_nil
+      expect(session[:carry_supplies]).to be_nil
```
Perhaps the test could set some values for `dietary_requirements` and `carry_supplies` before calling POST, so we can be sure that they're actually being cleared.

brucebolt replied:
Good spot!  I missed this fundamental part of the test.

leenagupte commented on app/controllers/coronavirus_form/essential_supplies_controller.rb:
```diff
@@ -18,12 +18,12 @@ def submit
       respond_to do |format|
         format.html { render controller_path, status: :unprocessable_entity }
       end
+    elsif session[:check_answers_seen] && session[:dietary_requirements].present? && session[:carry_supplies].present?
```
So the user can only go back to `check_answers` directly if they've already answered the `dietary_requirements ` and `carry_supplies` questions? What if they decide not to change their answer to `essential_supplies` and they submit a `yes` again?

brucebolt replied:
From the check answers page, they can go directly to the later questions if they want to change those questions, since as far as the user is aware these are individual questions rather than linked in some way.  I've only included this logic so they actually get asked the later questions if they are setting the answer to "Yes" for the first time.

leenagupte replied:
I think we've got some crossed wires. If the user previously answered "yes" to `essential_supplies`, completes the form (sees check answers), clicks to change their answer to `essential_supplies`, but then answers "yes" again, they should be routed back to check your answers, not `basic care needs`. Give it a go locally.

brucebolt replied:
Got it!  Was getting my yes and no responses mixed up, until I read the question properly 🤦.

leenagupte commented on spec/controllers/coronavirus_form/essential_supplies_controller_spec.rb:
```diff
@@ -57,8 +57,17 @@
+      expect(response).to redirect_to(dietary_requirements_path)
+    end
+
+    it "redirects to check your answers if check your answers previously seen and later food questions answered" do
```
Can you add a test for, `check_answers_seen` and later food questions not answered, and user answers "yes" to `essential_supplies`, as I think the behaviour needs to be different depending on the answer to that question.

brucebolt replied:
Added this.  Also picks up the missing logic you have raised in the controller.

leenagupte reviewed (commented):
This is looking really good. I just have a few concerns about the next question logic on the `essential_supplies` page.

I've tried this out locally, and if you answer "yes" to essential supplies the first time, you are routed to the basic needs page and then check your answers. Which is fine.
But if you click change your answers to essential supplies, and then submit "yes" again, you are still routed to the basic care needs page instead of the check your answers page.

brucebolt commented:
Thanks for the review @leenagupte.  I've addressed all your comments.

leenagupte commented:
@brucebolt You've got a merge conflict. Sorry about that! Also, I think the schema needs to be updated too.

brucebolt commented:
I fixed the merge conflict just as you commented 😁.

Will look at the schema after I've had lunch.

leenagupte commented on spec/controllers/coronavirus_form/essential_supplies_controller_spec.rb:
```diff
@@ -46,7 +50,10 @@
 
     it "clears session values for later food questions for a yes response" do
+      session[:dietary_requirements] = "Yes"
+      session[:carry_supplies] = "Yes"
```
Thanks for adding this. Minor nitpick is that this change should probably be fixed up.

brucebolt replied:
Ah, yes, didn't see that I'd added this test in an earlier commit.  Have fixed.

leenagupte commented on app/controllers/coronavirus_form/essential_supplies_controller.rb:
```diff
@@ -18,10 +18,13 @@ def submit
       update_session_store
       redirect_to dietary_requirements_url
-    elsif session[:check_answers_seen]
+    elsif session[:check_answers_seen] && @form_responses[:essential_supplies] == I18n.t("coronavirus_form.questions.essential_supplies.options.option_yes.label")
```
Thanks for fixing the logic. It has made the if/else a bit unwieldy though. Perhaps, rather than having this if/else, the logic for figuring out the next page could be moved to a method?

e.g.

```
  def next_page_url
    return dietary_requirements_url if answer_next_question?
    return check_your_answers_url if session[:check_answers_seen]

    basic_care_needs_url
  end

  def answer_next_question?
    @form_responses[:essential_supplies] == I18n.t("coronavirus_form.questions.essential_supplies.options.option_no.label") &&
      (session[:dietary_requirements].blank? || session[:carry_supplies].blank?)
  end
```

And then the routing logic could be simplified to:

```
    if invalid_fields.any?
      flash.now[:validation] = invalid_fields
      log_validation_error(invalid_fields)

      respond_to do |format|
        format.html { render controller_path, status: :unprocessable_entity }
      end
    else
      update_session_store
      redirect_to next_page_url
    end
```

brucebolt replied:
Nice idea, have made this change.

brucebolt committed:
Update schema to allow nil for later food questions

If the user does not answer the later food questions, we set their value
to be nil.  Therefore the schema needs updating to allow this.

leenagupte commented on spec/helpers/schema_helper_spec.rb:
```diff
@@ -274,6 +274,11 @@
         expect(validate_against_form_response_schema(data).first).to include("dietary_requirements")
       end
+
+      it "does not return a list of errors when carry_supplies has a nil value" do
```
Typo in test name. Should be `dietary_requirements`.

leenagupte replied:
Also, shouldn't the test on line 268: "returns a list of errors when dietary_requirements is missing" fail now?

That test should probably be replaced with something like:

```
      it "allows nhs_number to be blank" do
        data = valid_data.except(:nhs_number)
        expect(validate_against_form_response_schema(data)).to be_empty	        
      end
```

brucebolt replied:
We're not allowing the value to be missing, only to allow it to have a nil value.  Even if the user doesn't answer the later food questions, the value will always be nil: https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form/pull/358/files#diff-8c917c651898f2137b6f3b59a44e43f2R32-R33.

leenagupte commented on spec/helpers/schema_helper_spec.rb:
```diff
@@ -286,6 +291,11 @@
+      it "does not return a list of errors when carry_supplies has a nil value" do
+        data = valid_data.merge(carry_supplies: nil)
+        expect(validate_against_form_response_schema(data)).to be_empty
+      end
```
Again here, shouldn't the test on line 290, "returns a list of errors when carry_supplies has an unexpected value" fail as well?

leenagupte reviewed (commented):
Almost there. I just have a couple of comments about the schema tests.

leenagupte reviewed (approved):
💯 Looks great!
